File: Amazon/Trees and Graphs/Symmetric _tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(root == nullptr) return true;
        return check_sym(root->left, root->right);
    }
    
private:
    bool check_sym(TreeNode* l, TreeNode* r){
        if(l == nullptr and r == nullptr) return true;
        else if(l == nullptr or r == nullptr) return false;
        return ((l->val == r->val) and check_sym( l->left, r->right) and check_sym(l->right, r->left)) ;
    }
};

File: Amazon/Trees and Graphs/  Symmetric Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        return self.checkSym(root, root)
    
    def checkSym(self, l, r):
        if l is None and r is None: return True
        if l is None or r is None: return False
        
        return (l.val == r.val) and self.checkSym(l.left, r.right) and self.checkSym(l.right, r.left)